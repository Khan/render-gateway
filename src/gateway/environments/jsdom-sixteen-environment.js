// @flow
import type {ResourceLoader} from "jsdom";

import {applyAbortablePromisesPatch} from "./apply-abortable-promises-patch.js";

import type {IRenderEnvironment, RenderAPI, RenderResult} from "../types.js";

interface getResourceLoader {
    /**
     * Get a JSDOM resource loader for the given render request.
     *
     * @param {string} url The URL that is to be rendered. This is always
     * relative to the host and so does not contain protocol, hostname, nor port
     * information.
     * @param {RenderAPI} renderAPI An API of utilities for assisting with the
     * render operation.
     * @returns {ResourceLoader} A ResourceLoader instance for use with JSDOM.
     */
    (url: string, renderAPI: RenderAPI): ResourceLoader;
}

interface getFileList {
    /**
     * Get the list of file URLs to retrieve and execute for the given request.
     *
     * @param {string} url The URL that is to be rendered. This is always
     * relative to the host and so does not contain protocol, hostname, nor port
     * information.
     * @param {RenderAPI} renderAPI An API of utilities for assisting with the
     * render operation.
     * @returns {Promise<Array<string>>} An ordered array of absolute URLs for
     * the JavaScript files that are to be executed. These are exectued in the
     * same order as the array.
     */
    (url: string, renderAPI: RenderAPI): Promise<Array<string>>;
}

/**
 * A render environment built to support the JSDOM 16.x API.
 */
export class JSDOMSixteenEnvironment implements IRenderEnvironment {
    _getFileListFn: getFileList;
    _getResourceLoaderFn: getResourceLoader;

    /**
     * Create a new instance of this environment.
     *
     * @param {getFileList} getFileListFn
     * Callback that should return a promise for the list of JavaScript files
     * the environment must execute in order to produce a result for the given
     * render request.
     * @param {getResourceLoader} getResourceLoaderFn
     * Callback that should return a JSDOM resource loader for the given
     * request. We must call this per render so that logging is appropriately
     * channeled for the request being made.
     */
    constructor(
        getFileListFn: getFileList,
        getResourceLoaderFn: getResourceLoader,
    ) {
        if (getFileListFn == null) {
            throw new Error("Must provide callback for obtaining file list");
        }
        if (getResourceLoaderFn == null) {
            throw new Error(
                "Must provide callback for obtaining resource loader",
            );
        }
        this._getFileListFn = getFileListFn;
        this._getResourceLoaderFn = getResourceLoaderFn;

        applyAbortablePromisesPatch();
    }

    /**
     * Generate a render result for the given url.
     *
     * @param {string} url The URL that is to be rendered. This is always
     * relative to the host and so does not contain protocol, hostname, nor port
     * information.
     * @param {RenderAPI} renderAPI An API of utilities for assisting with the
     * render operation.
     * @returns {Promise<RenderResult>} The result of the render that is to be
     * returned by the gateway service as the response to the render request.
     * This includes the body of the response and the status code information.
     */
    render = async (
        url: string,
        renderAPI: RenderAPI,
    ): Promise<RenderResult> => {
        /**
         * First, we need to know what files to execute so that we can produce
         * a render result, and we need a resource loader so that we can
         * retrieve those files as well as support retrieving additional files
         * within our JSDOM environment.
         */
        const files = await this._getFileListFn(url, renderAPI);
        const resourceLoader = this._getResourceLoaderFn(url, renderAPI);

        /**
         * Now let's use the resource loader to get the files.
         * We ignore the `FetchOptions` param of resourceLoader.fetch as we
         * have nothing to pass there.
         */
        // eslint-disable-next-line no-unused-vars
        const packages = await Promise.all(
            files.map((f) => resourceLoader.fetch(f).then((b) => b.toString())),
        );

        /**
         * Right, we have the files. Now we need the JSDOM environment and the
         * rendering hooks so that we can make a render happen.
         */

        return Promise.resolve({
            body: "NOTHING",
            headers: {},
            status: 200,
        });
    };
}
