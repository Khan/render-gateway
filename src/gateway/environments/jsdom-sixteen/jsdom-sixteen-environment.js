// @flow
import type {ResourceLoader} from "jsdom";
import type {IJSDOMSixteenConfiguration} from "./index.js";
import type {IRenderEnvironment, RenderAPI, RenderResult} from "../../types.js";

/**
 * A render environment built to support the JSDOM 16.x API.
 */
export class JSDOMSixteenEnvironment implements IRenderEnvironment {
    _configuration: IJSDOMSixteenConfiguration;

    /**
     * Create a new instance of this environment.
     *
     * @param {IJSDOMSixteenConfiguration} configuration
     * Configuration for the environment.
     */
    constructor(configuration: IJSDOMSixteenConfiguration) {
        if (configuration == null) {
            throw new Error("Must provide environment configuration");
        }
        this._configuration = configuration;
    }

    _retrieveTargetFiles = async (
        url: string,
        renderAPI: RenderAPI,
        resourceLoader: ResourceLoader,
    ): Promise<Array<string>> => {
        const traceSession = renderAPI.trace("Retrieving target files");
        try {
            /**
             * First, we need to know what files to execute so that we can produce
             * a render result, and we need a resource loader so that we can
             * retrieve those files as well as support retrieving additional files
             * within our JSDOM environment.
             */
            const files = await this._configuration.getFileList(url, renderAPI);

            /**
             * Now let's use the resource loader to get the files.
             * We ignore the `FetchOptions` param of resourceLoader.fetch as we
             * have nothing to pass there.
             */
            return await Promise.all(
                files.map((f) => {
                    const fetchResult = resourceLoader.fetch(f);
                    /**
                     * Resource loader's fetch can return null. It shouldn't for
                     * any of these files though, so if it does, let's raise an
                     * error!
                     */
                    if (fetchResult == null) {
                        throw new Error(
                            `Unable to retrieve ${f}. ResourceLoader returned null.`,
                        );
                    }
                    /**
                     * No need to reconnect the abort() in this case since we
                     * won't be calling it.
                     */
                    return fetchResult.then((b) => b.toString());
                }),
            );
        } finally {
            traceSession.end();
        }
    };

    /**
     * Generate a render result for the given url.
     *
     * @param {string} url The URL that is to be rendered. This is always
     * relative to the host and so does not contain protocol, hostname, nor port
     * information.
     * @param {RenderAPI} renderAPI An API of utilities for assisting with the
     * render operation.
     * @returns {Promise<RenderResult>} The result of the render that is to be
     * returned by the gateway service as the response to the render request.
     * This includes the body of the response and the status code information.
     */
    render = async (
        url: string,
        renderAPI: RenderAPI,
    ): Promise<RenderResult> => {
        /**
         * We are going to need a resource loader so that we can obtain files
         * both inside and outside the JSDOM VM.
         */
        const resourceLoader = this._configuration.getResourceLoader(
            url,
            renderAPI,
        );

        // eslint-disable-next-line no-unused-vars
        const files = await this._retrieveTargetFiles(
            url,
            renderAPI,
            resourceLoader,
        );

        /**
         * Right, we have the files. Now we need the JSDOM environment and the
         * rendering hooks so that we can make a render happen.
         */
        /**
         * TODO(somewhatabstract): All the things
         * 1. Need to setup the JSDOM VM
         *    - see createRenderContext for RRS
         *    - call the afterEnvSetup and attach anything it returns to the
         *      vm context
         *
         * 2. Need to setup the environment with render registration callbacks
         *    - see the render function in render.js of RRS
         *
         * 3. Invoke the render and return its result.
         *
         * 4. Finally, close the JSDOM environment.
         */

        return Promise.resolve({
            body: "NOTHING",
            headers: {},
            status: 200,
        });
    };
}
