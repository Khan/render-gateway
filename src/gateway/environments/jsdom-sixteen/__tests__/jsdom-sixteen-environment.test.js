// @flow
import {JSDOMSixteenEnvironment} from "../jsdom-sixteen-environment.js";

describe("JSDOMSixteenEnvironment", () => {
    describe("#constructor", () => {
        it("should throw if configuration is not provided", () => {
            // Arrange

            // Act
            const underTest = () => new JSDOMSixteenEnvironment((null: any));

            // Assert
            expect(underTest).toThrowErrorMatchingInlineSnapshot(
                `"Must provide environment configuration"`,
            );
        });
    });

    describe("#render", () => {
        it("should get a resource loader", async () => {
            // Arrange
            const fakeLogger: any = "FAKE_LOGGER";
            const fakeTraceSession: any = {
                end: jest.fn(),
                addLabel: jest.fn(),
            };
            const fakeRenderAPI: any = {
                trace: jest.fn().mockReturnValue(fakeTraceSession),
                getHeader: jest.fn(),
                logger: fakeLogger,
            };
            const fakeConfiguration = {
                getFileList: jest.fn().mockReturnValue(Promise.resolve([])),
                getResourceLoader: jest.fn(),
                afterEnvSetup: jest.fn(),
            };
            const underTest = new JSDOMSixteenEnvironment(fakeConfiguration);

            // Act
            await underTest.render("URL", fakeRenderAPI);

            // Assert
            expect(fakeConfiguration.getResourceLoader).toHaveBeenCalledWith(
                "URL",
                fakeRenderAPI,
            );
        });

        it("should retrieve the list of files to be downloaded", async () => {
            // Arrange
            const fakeLogger: any = "FAKE_LOGGER";
            const fakeTraceSession: any = {
                end: jest.fn(),
                addLabel: jest.fn(),
            };
            const fakeRenderAPI: any = {
                trace: jest.fn().mockReturnValue(fakeTraceSession),
                getHeader: jest.fn(),
                logger: fakeLogger,
            };
            const fakeConfiguration = {
                getFileList: jest.fn().mockReturnValue(Promise.resolve([])),
                getResourceLoader: jest.fn(),
                afterEnvSetup: jest.fn(),
            };
            const underTest = new JSDOMSixteenEnvironment(fakeConfiguration);

            // Act
            await underTest.render("URL", fakeRenderAPI);

            // Assert
            expect(fakeConfiguration.getFileList).toHaveBeenCalledWith(
                "URL",
                fakeRenderAPI,
            );
        });

        it("should trace the file acquisition phase", async () => {
            // Arrange
            const fakeLogger: any = "FAKE_LOGGER";
            const fakeTraceSession: any = {
                end: jest.fn(),
                addLabel: jest.fn(),
            };
            const fakeRenderAPI: any = {
                trace: jest.fn().mockReturnValue(fakeTraceSession),
                getHeader: jest.fn(),
                logger: fakeLogger,
            };
            const fakeConfiguration = {
                getFileList: jest.fn().mockReturnValue(Promise.resolve([])),
                getResourceLoader: jest.fn(),
                afterEnvSetup: jest.fn(),
            };
            const underTest = new JSDOMSixteenEnvironment(fakeConfiguration);

            // Act
            await underTest.render("URL", fakeRenderAPI);

            // Assert
            expect(fakeRenderAPI.trace).toHaveBeenCalledBefore(
                fakeConfiguration.getFileList,
            );
        });

        it("should fetch the files via the resource loader", async () => {
            // Arrange
            const fakeLogger: any = "FAKE_LOGGER";
            const fakeTraceSession: any = {
                end: jest.fn(),
                addLabel: jest.fn(),
            };
            const fakeRenderAPI: any = {
                trace: jest.fn().mockReturnValue(fakeTraceSession),
                getHeader: jest.fn(),
                logger: fakeLogger,
            };
            const fakeResourceLoader: any = {
                fetch: jest
                    .fn()
                    .mockImplementation((f) =>
                        Promise.resolve(`FETCHED: ${f}`),
                    ),
            };
            const fakeConfiguration = {
                getFileList: jest
                    .fn()
                    .mockReturnValue(
                        Promise.resolve(["filea", "fileb", "filec"]),
                    ),
                getResourceLoader: jest
                    .fn()
                    .mockReturnValue(fakeResourceLoader),
                afterEnvSetup: jest.fn(),
            };
            const underTest = new JSDOMSixteenEnvironment(fakeConfiguration);

            // Act
            await underTest.render("URL", fakeRenderAPI);

            // Assert
            expect(fakeResourceLoader.fetch).toHaveBeenCalledWith("filea");
            expect(fakeResourceLoader.fetch).toHaveBeenCalledWith("fileb");
            expect(fakeResourceLoader.fetch).toHaveBeenCalledWith("filec");
        });

        it("should throw if file fetch returns null", async () => {
            // Arrange
            const fakeLogger: any = "FAKE_LOGGER";
            const fakeTraceSession: any = {
                end: jest.fn(),
                addLabel: jest.fn(),
            };
            const fakeRenderAPI: any = {
                trace: jest.fn().mockReturnValue(fakeTraceSession),
                getHeader: jest.fn(),
                logger: fakeLogger,
            };
            const fakeResourceLoader: any = {
                fetch: jest.fn().mockReturnValue(null),
            };
            const fakeConfiguration = {
                getFileList: jest
                    .fn()
                    .mockReturnValue(Promise.resolve(["BAD_FILE"])),
                getResourceLoader: jest
                    .fn()
                    .mockReturnValue(fakeResourceLoader),
                afterEnvSetup: jest.fn(),
            };
            const env = new JSDOMSixteenEnvironment(fakeConfiguration);

            // Act
            const underTest = env.render("URL", fakeRenderAPI);

            // Assert
            await expect(underTest).rejects.toThrowErrorMatchingInlineSnapshot(
                `"Unable to retrieve BAD_FILE. ResourceLoader returned null."`,
            );
        });

        it("should end the trace session if file acquisition throws", async () => {
            // Arrange
            const fakeLogger: any = "FAKE_LOGGER";
            const fakeTraceSession: any = {
                end: jest.fn(),
                addLabel: jest.fn(),
            };
            const fakeRenderAPI: any = {
                trace: jest.fn().mockReturnValue(fakeTraceSession),
                getHeader: jest.fn(),
                logger: fakeLogger,
            };
            const fakeResourceLoader: any = {
                fetch: jest.fn().mockReturnValue(Promise.resolve(null)),
            };
            const fakeConfiguration = {
                getFileList: jest
                    .fn()
                    .mockReturnValue(Promise.resolve(["BAD_FILE"])),
                getResourceLoader: jest
                    .fn()
                    .mockReturnValue(fakeResourceLoader),
                afterEnvSetup: jest.fn(),
            };
            const env = new JSDOMSixteenEnvironment(fakeConfiguration);

            // Act
            await env.render("URL", fakeRenderAPI).catch(() => {
                /* NOTHING */
            });

            // Assert
            expect(fakeTraceSession.end).toHaveBeenCalled();
        });

        it("should end the trace session if file acquisition succeeds", async () => {
            // Arrange
            const fakeLogger: any = "FAKE_LOGGER";
            const fakeTraceSession: any = {
                end: jest.fn(),
                addLabel: jest.fn(),
            };
            const fakeRenderAPI: any = {
                trace: jest.fn().mockReturnValue(fakeTraceSession),
                getHeader: jest.fn(),
                logger: fakeLogger,
            };
            const fakeResourceLoader: any = {
                fetch: jest
                    .fn()
                    .mockImplementation((f) =>
                        Promise.resolve(`FETCHED: ${f}`),
                    ),
            };
            const fakeConfiguration = {
                getFileList: jest
                    .fn()
                    .mockReturnValue(
                        Promise.resolve(["filea", "fileb", "filec"]),
                    ),
                getResourceLoader: jest
                    .fn()
                    .mockReturnValue(fakeResourceLoader),
                afterEnvSetup: jest.fn(),
            };
            const env = new JSDOMSixteenEnvironment(fakeConfiguration);

            // Act
            await env.render("URL", fakeRenderAPI);

            // Assert
            expect(fakeTraceSession.end).toHaveBeenCalled();
        });

        it.todo("should setup the JSDOM VM environment");

        it.todo("should apply afterEnvSetup");

        it.todo("should invoke the render");

        it.todo("should return the render result");
    });
});
