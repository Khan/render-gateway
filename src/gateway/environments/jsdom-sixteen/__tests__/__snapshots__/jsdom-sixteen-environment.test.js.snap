// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSDOMSixteenEnvironment #render should reject if result is malformed 1`] = `"Malformed render result: undefined"`;

exports[`JSDOMSixteenEnvironment #render should reject if result is malformed 2`] = `"Malformed render result: null"`;

exports[`JSDOMSixteenEnvironment #render should reject if result is malformed 3`] = `"Malformed render result: \\"THIS IS NOT CORRECT\\""`;

exports[`JSDOMSixteenEnvironment #render should reject if result is malformed 4`] = `"Malformed render result: {\\"status\\":200,\\"headers\\":{}}"`;

exports[`JSDOMSixteenEnvironment #render should reject if result is malformed 5`] = `"Malformed render result: {\\"body\\":\\"NEED MORE THAN THIS\\"}"`;

exports[`JSDOMSixteenEnvironment #render should reject if result is malformed 6`] = `"Malformed render result: {\\"body\\":\\"THIS HELPS BUT WHERE ARE THE HEADERS\\",\\"status\\":200}"`;
